src/operator/nn/softmax-inl.h

```diff
@@ -177,6 +231,51 @@ inline void SoftmaxGrad(Stream<cpu> *s, OType *out, OType *ograd,
   }
 }
 
+template<typename OP1, typename OP2, int Req, bool negate,
+  typename AType, typename DType, typename OType, typename IType, int ndim>
+inline void SoftmaxWithLengthGrad(Stream<cpu> *s, OType *out, OType *ograd,
+                                  DType *igrad, IType *length, Shape<ndim> shape,
+                                  int axis, const DType temperature) {
+  index_t M = shape[axis];
+  index_t N = shape.Size()/M;
+  Shape<ndim> stride = calc_stride(shape);
+  Shape<ndim> sshape = shape;
+  sshape[axis] = 1;
+  index_t sa = stride[axis];
+
+  #pragma omp parallel for
+  for (int i = 0; i < static_cast<int>(N); ++i) {
+    index_t base = unravel_dot(i, sshape, stride);
+    IType len = length[i];
```

Comment:
Can we simply do something like
```suggestion
    index_t len = static_cast<index_t>(length[i]);
```
Shouldn't the data type of elements inside length tensor always be the same as `M`?

Reply:
Why does it have to be the same data type as `M`? It's a tensor provided by the user so the user could be providing a tensor of any data type.